near-contract-standards/src/non_fungible_token/core_impl.rs

```diff
@@ -0,0 +1,442 @@
+                );
+            }
+        }
+
+        assert_ne!(&owner_id, receiver_id, "Current and next owner must differ");
+
+        self.internal_transfer_unguarded(&token_id, &owner_id, &receiver_id);
+
+        log!("Transfer {} from {} to {}", token_id, sender_id, receiver_id);
+        if let Some(memo) = memo {
+            log!("Memo: {}", memo);
+        }
+
+        // return previous owner & approvals
+        (owner_id, approved_account_ids)
+    }
+}
+
+impl NonFungibleTokenCore for NonFungibleToken {
+    fn nft_transfer(
+        &mut self,
+        receiver_id: ValidAccountId,
+        token_id: TokenId,
+        approval_id: Option<u64>,
+        memo: Option<String>,
+    ) {
+        assert_one_yocto();
+        let sender_id = env::predecessor_account_id();
+        self.internal_transfer(&sender_id, receiver_id.as_ref(), &token_id, approval_id, memo);
+    }
+
+    fn nft_transfer_call(
+        &mut self,
+        receiver_id: ValidAccountId,
+        token_id: TokenId,
+        approval_id: Option<u64>,
+        memo: Option<String>,
+        msg: String,
+    ) -> PromiseOrValue<bool> {
+        assert_one_yocto();
+        let sender_id = env::predecessor_account_id();
+        let (old_owner, old_approvals) =
+            self.internal_transfer(&sender_id, receiver_id.as_ref(), &token_id, approval_id, memo);
+        // Initiating receiver's call and the callback
+        ext_receiver::nft_on_transfer(
+            sender_id.clone(),
+            old_owner.clone(),
+            token_id.clone(),
+            msg,
+            receiver_id.as_ref(),
+            NO_DEPOSIT,
+            env::prepaid_gas() - GAS_FOR_FT_TRANSFER_CALL,
+        )
+        .then(ext_self::nft_resolve_transfer(
+            old_owner,
+            receiver_id.into(),
+            token_id,
+            old_approvals,
+            &env::current_account_id(),
+            NO_DEPOSIT,
+            GAS_FOR_RESOLVE_TRANSFER,
+        ))
+        .into()
+    }
+
+    fn nft_token(self, token_id: TokenId) -> Option<Token> {
+        let owner_id = self.owner_by_id.get(&token_id)?;
+        let metadata = self.token_metadata_by_id.and_then(|by_id| by_id.get(&token_id));
+        let approved_account_ids = self
+            .approvals_by_id
+            .and_then(|by_id| by_id.get(&token_id).or_else(|| Some(HashMap::new())));
+        Some(Token { token_id, owner_id, metadata, approved_account_ids })
+    }
+
+    fn mint(
+        &mut self,
+        token_id: TokenId,
+        token_owner_id: ValidAccountId,
+        token_metadata: Option<TokenMetadata>,
+    ) -> Token {
+        // TODO: charge someone for storage
+        assert_at_least_one_yocto();
+        assert_eq!(env::predecessor_account_id(), self.owner_id, "Unauthorized");
+        if self.token_metadata_by_id.is_some() && token_metadata.is_none() {
+            env::panic(b"Must provide metadata");
+        }
+        if let Some(_) = self.owner_by_id.get(&token_id) {
```

Comment:
```suggestion
        if self.owner_by_id.get(&token_id).is_some() {
```
Seems like checking key existence should be a function on unordered_map, this is doing an unnecessary read

Reply:
This was lost during the refactor to TreeMap

Reply:
yeah that's fair, but there is still no need for the `if let Some(_)` pattern here, so I think my suggestion is still valid